xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/skin/AbstractWikiResource.java

```diff
@@ -91,4 +94,21 @@ public String getURL(boolean forceSkinAction) throws Exception
     }
 
     public abstract String getURL(XWikiDocument document) throws Exception;
+
+    @Override
+    public String getURL(boolean forceSkinAction, Map<String, String> queryParameters) throws Exception
+    {
+        return getURL(getDocument());
+    }
+
+    /**
+     * Create URL for the given document with some query parameters.
+     * @param document the document wanted for the URL
+     * @param queryParameters some parameters to add in the URL
+     * @return a URL for the given document with some parameters
+     * @throws Exception if the URL cannot be built.
+     * @since 11.0RC1
+     */
+    @Unstable
+    public abstract String getURL(XWikiDocument document, Map<String, String> queryParameters) throws Exception;
```

Comment:
Looks weird to pass query parameters to get them back after the call. @tmortagne could you review this API to verify the design is good?

Reply:
Not sure what you mean by "get them back after the call". What you get back is a URL.

From what I understand the point of this API is to provide custom parameters to add to the generated URL. You have the same thing in XWikiURLFactory.